coderd/workspaceagents.go

```diff
@@ -755,31 +755,69 @@ func convertWorkspaceAgent(derpMap *tailcfg.DERPMap, coordinator tailnet.Coordin
 func (api *API) workspaceAgentReportStats(rw http.ResponseWriter, r *http.Request) {
 	ctx := r.Context()
 
-	api.WebsocketWaitMutex.Lock()
-	api.WebsocketWaitGroup.Add(1)
-	api.WebsocketWaitMutex.Unlock()
-	defer api.WebsocketWaitGroup.Done()
-
 	workspaceAgent := httpmw.WorkspaceAgent(r)
-	resource, err := api.Database.GetWorkspaceResourceByID(ctx, workspaceAgent.ResourceID)
+	workspace, err := api.Database.GetWorkspaceByAgentID(ctx, workspaceAgent.ID)
 	if err != nil {
 		httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
-			Message: "Failed to get workspace resource.",
+			Message: "Failed to get workspace.",
 			Detail:  err.Error(),
 		})
 		return
 	}
 
-	build, err := api.Database.GetWorkspaceBuildByJobID(ctx, resource.JobID)
-	if err != nil {
-		httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
-			Message: "Failed to get build.",
-			Detail:  err.Error(),
+	var req codersdk.AgentStats
+	if !httpapi.Read(ctx, rw, r, &req) {
+		return
+	}
+
+	if req.RxBytes == 0 && req.TxBytes == 0 {
+		httpapi.Write(ctx, rw, http.StatusOK, codersdk.AgentStatsResponse{
+			ReportInterval: api.AgentStatsRefreshInterval,
 		})
 		return
 	}
 
-	workspace, err := api.Database.GetWorkspaceByID(ctx, build.WorkspaceID)
+	activityBumpWorkspace(api.Logger.Named("activity_bump"), api.Database, workspace.ID)
+
+	now := database.Now()
+	_, err = api.Database.InsertAgentStat(ctx, database.InsertAgentStatParams{
+		ID:          uuid.New(),
+		CreatedAt:   now,
+		AgentID:     workspaceAgent.ID,
+		WorkspaceID: workspace.ID,
+		UserID:      workspace.OwnerID,
+		TemplateID:  workspace.TemplateID,
+		Payload:     json.RawMessage("{}"),
+	})
+	if err != nil {
+		httpapi.InternalServerError(rw, err)
+		return
+	}
+
+	err = api.Database.UpdateWorkspaceLastUsedAt(ctx, database.UpdateWorkspaceLastUsedAtParams{
+		ID:         workspace.ID,
+		LastUsedAt: now,
```

Comment:
would this not cause every workspace to get bumped every 5 minutes regardless of activity as long as it had ever received and transmitted at least 1 byte?

Reply:
Oh, do the counters get reset each push?

Reply:
Yeah the counters get reset on each push, so you're only sending the activity from that period.